src/daft-core/src/array/ops/hash.rs

```diff
@@ -88,6 +92,103 @@ impl NullArray {
+    // first we hash the flat child
+    // turning [[stuff], [stuff, stuff], ...] into [[hash], [hash, hash], ...]
+    // then we hash each sublist as bytes, giving us [hash, hash, ...] as desired
+    // if seed is provided, the sublists are hashed with the seed broadcasted
+
+    if let Some(seed_arr) = seed {
+        let combined_validity = arrow_bitmap_and_helper(validity, seed.unwrap().validity());
+        UInt64Array::from_iter(
+            name,
+            u64::range(0, offsets.len() - 1).unwrap().map(|i| {
+                let start = offsets[i as usize] as usize;
+                let end = offsets[i as usize + 1] as usize;
+                // apply the current seed across this row
+                let cur_seed_opt = seed_arr.get(i as usize);
+                let flat_seed = UInt64Array::from_iter(
+                    "seed",
+                    std::iter::repeat(cur_seed_opt).take(end - start),
+                );
+                let hashed_child = flat_child
+                    .slice(start, end)
+                    .ok()?
+                    .hash(Some(&flat_seed))
+                    .ok()?;
+                let child_bytes: Vec<u8> = hashed_child
+                    .as_arrow()
+                    .values_iter()
+                    .flat_map(|v| v.to_le_bytes())
+                    .collect();
+                if let Some(cur_seed) = cur_seed_opt {
+                    Some(xxh3_64_with_seed(&child_bytes, cur_seed))
+                } else {
+                    Some(xxh3_64(&child_bytes))
+                }
+            }),
+        )
+        .with_validity(combined_validity)
+    } else {
+        // since we don't have a seed we can hash entire flat child at once
+        let hashed_child = flat_child.hash(None)?;
+        // hashing collects the array anyways so this collect doesn't matter
+        let child_bytes: Vec<u8> = hashed_child
+            .as_arrow()
+            .values_iter()
+            .flat_map(|v| v.to_le_bytes())
+            .collect();
+        const OFFSET: usize = (u64::BITS as usize) / 8; // how many bytes per u64
+        let combined_validity = validity.cloned();
+        UInt64Array::from_iter(
+            name,
+            u64::range(0, offsets.len() - 1).unwrap().map(|i| {
+                let start = (offsets[i as usize] as usize) * OFFSET;
+                let end = (offsets[i as usize + 1] as usize) * OFFSET;
+                Some(xxh3_64(&child_bytes[start..end]))
+            }),
+        )
+        .with_validity(combined_validity)
+    }
+}
+
+impl ListArray {
+    pub fn hash(&self, seed: Option<&UInt64Array>) -> DaftResult<UInt64Array> {
+        hash_list(
+            self.name(),
+            self.offsets(),
+            &self.flat_child,
+            self.validity(),
+            seed,
+        )
+    }
+}
+
+impl FixedSizeListArray {
+    pub fn hash(&self, seed: Option<&UInt64Array>) -> DaftResult<UInt64Array> {
+        let size = match self.field.dtype {
```

Comment:
nit: you should be able to use `self.size()` -- we duplicate the size information so users don't have to match on the field like this!